File: Algorithm/Python/957. Prison Cells After N Days.py
# -*- coding: utf-8 -*-
# Created on Sun Dec 16 2018 10:54:0
# Author: WuLC
# EMail: liangchaowu5@gmail.com

# hashtable
# there is always a loop, but how to prove
class Solution(object):
    def prisonAfterNDays(self, cells, N):
        """
        :type cells: List[int]
        :type N: int
        :rtype: List[int]
        """
        p1, p2 = cells, [0] * len(cells)
        record, reversed_record = {}, {}
        for i in xrange(N):
            num = reduce(lambda x, y: str(x)+str(y), p1)
            if num in record:
                left = (N - record[num]) % (i - record[num])
                return [int(c) for c in reversed_record[record[num] + left]]
            else:
                record[num] =  i
                reversed_record[i] = num
            for j in xrange(1, len(cells) - 1):
                p2[j] = 1 if p1[j-1] == p1[j+1] else 0
            p1, p2 = p2, p1
            p2[0], p2[-1] = 0, 0
        return p1

File: Algorithm/C++/957. Prison Cells After N Days.cc
/*
 * Created on Sun Dec 16 2018 20:19:59
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */

// hashmap, there always exists a loop 
#include <vector>
#include <unordered_map>
#include <string>

using std::vector;
using std::unordered_map;
using std::string;

class Solution {
 public:
  vector<int> prisonAfterNDays(vector<int>& cells, int N) {
    unordered_map<string, int> record;
    unordered_map<int, vector<int>> reversed_record;
    vector<int> curr(cells);
    for (int i = 0; i < N; i++) {
      std::stringstream ss;
      for(auto digit:curr) ss << char(digit);
      string s = ss.str();
      if (record.count(s) > 0) {
        int left = (N - record[s]) % (i - record[s]);
        return reversed_record[record[s] + left];
      } else {
        record[s] = i;
        reversed_record[i] = curr;
      }

      // iterate
      vector<int> next(8, 0);
      for (int j = 1; j < 7; j++)
        next[j] = curr[j-1] == curr[j+1]? 1: 0;
      curr = next;
    }
    return curr;
  }
};